File: projects/order-creation/app/src/main/java/com/robinfood/app/controllers/transactions/ITransactionController.java
package com.robinfood.app.controllers.transactions;

import com.robinfood.core.dtos.apiresponsebuilder.ApiResponseDTO;
import com.robinfood.core.dtos.transactionrequestdto.TransactionRequestDTO;
import com.robinfood.core.dtos.transactionresponsedto.TransactionCreationResponseDTO;
import com.robinfood.core.exceptions.TransactionCreationException;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;

/**
 * Exposes the API that handles all data related to transactions
 */
@Tag(name = "Transactions", description = "Requests for transactions related data")
public interface ITransactionController {

    /**
     * Validates all the necessary steps to create a transaction and then creates it
     *
     * @return the details of the created transaction and its orders
     */
    @Operation(summary = "Validates all the necessary steps to create a transaction and then creates it")
    @ApiResponses(value = {
            @ApiResponse(
                    responseCode = "201",
                    description = "Returns the details of the created transaction and its orders",
                    content = {
                            @Content(mediaType = MediaType.APPLICATION_JSON_VALUE,
                                    schema = @Schema(implementation = TransactionCreationResponseDTO.class))
                    })
    })
    ResponseEntity<ApiResponseDTO<TransactionCreationResponseDTO>> createTransaction(
            TransactionRequestDTO transactionRequest
    ) throws TransactionCreationException;
}

File: projects/posv2/app/src/main/kotlin/com/robinfood/app/controllers/transaction/ITransactionController.kt
package com.robinfood.app.controllers.transaction

import com.robinfood.core.dtos.ApiResponseDTO
import com.robinfood.core.dtos.transactionrequest.TransactionRequestDTO
import com.robinfood.core.dtos.transactionresponse.TransactionCreatedResponseDTO
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.media.ArraySchema
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import javax.servlet.http.HttpServletRequest

/**
 * Exposes the API that handles all data related to transactions
 */
@Tag(name = "Transactions", description = "Requests for transactions related data")
interface ITransactionController {

    /**
     * Sends request to create a transaction
     * @return The order object after being created
     */
    @Operation(summary = "Sends a request to create a transaction")
    @ApiResponses(
        ApiResponse(
            responseCode = "201",
            description = "This sends a request to create a transaction",
            content = [(Content(
                array = ArraySchema(schema = Schema(implementation = TransactionCreatedResponseDTO::class)),
                mediaType = MediaType.APPLICATION_JSON_VALUE
            ))]
        )
    )
    suspend fun createTransaction(
        httpServletRequest: HttpServletRequest,
        transactionRequestDTO: TransactionRequestDTO
    ): ResponseEntity<ApiResponseDTO<TransactionCreatedResponseDTO>>
}